firedrake/mesh.py

```diff
@@ -3162,57 +3180,40 @@ def _pic_swarm_in_mesh(
-    remote_ranks = owned_ranks_local_visible[local_halo_idxs]
-
-    # Remote index is the offset in memory of the point on the other rank.
-    # Finding them is a bit of a faff...
-    remote_idxs = np.empty(npts_halo, dtype=IntType)
-    # global indices are unique particle identifiers - where they turn up on
-    # other ranks, they refer to the same particle.
-    global_idxs_local_visible = global_idxs_local[visible_idxs]
-    global_idxs_local_visible_halo = global_idxs_local_visible[local_halo_idxs]
-    global_idxs_local_visible_owned = global_idxs_local_visible[is_owned]
-    global_idxs_local_visible_owned_allranks = parent_mesh.comm.allgather(
-        global_idxs_local_visible_owned
-    )
-    # global_idxs_local_visible_owned_allranks is in rank order, so we can
-    # index into it with rank number (as long as our ranks start from zero!).
-    # We can now search for global index matches on each remote rank.
-    unique_remote_ranks = np.unique(remote_ranks)  # Don't need one for each halo point!
-    for unique_remote_rank in unique_remote_ranks:
-        remote_idxs[remote_ranks == unique_remote_rank] = np.where(
-            np.in1d(
-                global_idxs_local_visible_owned_allranks[unique_remote_rank],
-                global_idxs_local_visible_halo,
-            )
-        )[0]
-
-    # Interleave each rank and index into (rank, index) pairs for use as remote
-    # in the SF
-    remote_ranks_and_idxs = np.empty(2*nleaves, dtype=IntType)
-    remote_ranks_and_idxs[0::2] = remote_ranks
-    remote_ranks_and_idxs[1::2] = remote_idxs
-    sf = swarm.getPointSF()
-    sf.setGraph(nroots, local_halo_idxs, remote_ranks_and_idxs)
-    swarm.setPointSF(sf)
-
     # Note when getting original ordering for extruded meshes we recalculate
     # the base_parent_cell_nums and extrusion_heights - note this could
     # be an SF operation
     if redundant and parent_mesh.comm.rank != 0:
         original_ordering_swarm_coords = np.empty(shape=(0, coords.shape[1]))
     else:
         original_ordering_swarm_coords = coords
+    # Set pointSF
+    owners = swarm.getField("DMSwarm_rank")
+    swarm.restoreField("DMSwarm_rank")
+    halo_indices, = np.where(owners != parent_mesh.comm.rank)
+    halo_indices = halo_indices.astype(IntType)
+    n = coords.shape[0]
+    m = owners.shape[0]
+    _input_ordering_sf = VertexOnlyMeshTopology._make_input_ordering_sf(swarm, n, None)
+    a = np.empty(n, dtype=utils.IntType)
+    a.fill(-1)
+    b = np.arange(m, dtype=utils.IntType)  # swarm local point numbers
+    b[halo_indices] = -1
+    unit = MPI._typedict[np.dtype(utils.IntType).char]
+    _input_ordering_sf.reduceBegin(unit, b, a, MPI.MAX)
+    _input_ordering_sf.reduceEnd(unit, b, a, MPI.MAX)
+    _input_ordering_sf.bcastBegin(unit, a, b, MPI.REPLACE)
+    _input_ordering_sf.bcastEnd(unit, a, b, MPI.REPLACE)
+    if np.any(b < 0):
+        raise RuntimeError("Unable to make swarm pointSF due to inconsistent data")
```

Comment:
Clever! Can you explain what's going on here? I think I understand that the input ordering SF should has enough information to tell you the remote ranks and indices, but at first glance it's not quite obvious how you're getting 
them.

Using the input ordering SF to reduce the swarm local point numbers `b` into an array of integers (of length equal to the number of supplied coordinates that were rearranged or dropped)  `a`, with `MPI.MAX` does... what?

And then you broadcast in the other direction which somehow gives you all the local indices?

What does unit do?

Reply:
Yes. All points in `b` are leaves. Then, `b[halo_indices] = -1`, `a.fill(-1)`, and `MPI.MAX` ensure that `a` has the swarm local point numbers of the owners after `reduce`. `a[j] = -1` if `j` corresponds to a missing point. Then, `broadcast` updates `b[halo_indices]` (it also updates b[~halo_indices]`, not changing any values there).

Reply:
Please explain the algorithm in comments, as you have here. 

I'm also keen on more descriptive names which fit with the rest of the function. If I've understood you correctly I think `b` is `owner_local_idx`. I'm not sure of what to call `a` though - is it `remote_local_idx`?

Reply:
Added. With added comment, I do not think we need longer variable names. Their life time is only 10 lines.

Reply:
`a->local_points_reduced`
`b->local_points`
What do you find difficult to understand? The best way to convince yourself would be to run a simple parallel example and check data line by line.